editor.planx.uk/src/pages/FlowEditor/lib/store/editor.ts

```diff
@@ -437,4 +441,9 @@ export const editorStore: StateCreator<
     })(get().flow);
     send(ops);
   },
+
+  undoOperation: (op: Operation) => {
+    const inverseOps: OT.Op[] = type.invert(op.data);
+    send(inverseOps);
+  },
```

Comment:
I considered LOTS of different ways of doing this piece - but eventually landed on this simple Editor store method which hooks into our existing `send(ops)` - meaning we don't have to directly write any mutations to update `operations` or `flows` as the original script does: https://github.com/theopensystemslab/planx-data-recovery/blob/main/index.js

The other main route I was considering was creating an API endpoint to mimic the full set of data-recovery steps (eg `/flows/:flowId/undo/:operationId`). Ultimately decided against this though in favor of an editor-only approach for a couple key reasons: 
- the API currently knows very little (nothing?) about operations & ShareDB which makes it hard to access/reuse types, existing methods, etc. Handling "undo" in the Editor Store means this logic is co-located with other graph operations like add/remove/etc. 
- the original script has a lot of safety nets (eg print original flow data, proposed flow data, diff the JSONs, etc) which we don't actually need to replicate here for an MVP. It would be really nice to figure out an intermediary "ConfirmationDialog" step here in the Editor, but that can be a nice-to-have followup (it's actually sort of cool how seamless it feels as-is!)

Reply:
Super helpful explanation here thanks.